### PR TITLE
[WIP] MaprouletteConfiguration - Adding a new constructor to handle connection without project name.

### PR DESCRIPTION
**Pull request by smaheshwaram**
### Description:
This PR addresses the Maproulette connection issue without the need of project name in the configuration.

### Unit Test Approach:

Tested all the existing unit tests and also fixed one of the failing unit test related to updating a task.
Added unit tests necessary for new configuration.

### Test Results:

It is working as expected. All unit tests ran successfully.

------

In doubt: [Contributing Guidelines](../CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/maproulette/client/connection/MapRouletteConfiguration.java [hunk 1]
```diff
@@ -33,7 +33,7 @@
     private final int port;
     private final String scheme;
     private final String server;
-    private final String defaultProjectName;
+    private String defaultProjectName = "";
```
Comment (mgcuthbert) [inline src/main/java/org/maproulette/client/connection/MapRouletteConfiguration.java:36]: This should remain final, and you can set the value in the constructor to an empty string if you want.

File: src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ public void updateTest() throws MapRouletteException
         this.compare(toCreate, created.get());
 
         final var update = created.get().toBuilder().instruction("UpdatedInstruction")
-                .status(TaskStatus.FIXED).name("UpdateTaskName")
+                .status(TaskStatus.FIXED).name("UpdateTaskName").resetGeometry()
```
Comment (mgcuthbert) [inline src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java:67]: Why is this function call being added to the test?

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java [hunk 3]
```diff
@@ -95,7 +102,7 @@ public MapRouletteConfiguration getConfiguration()
     public void setup() throws MapRouletteException
     {
         // build the project that will be used to execute the integration tests for the challenges
-        this.defaultProjectIdentifier = this.projectAPI.create(this.defaultProject).getId();
+        this.defaultProjectIdentifier = getProjectAPI().create(this.defaultProject).getId();
```
Comment (mgcuthbert) [inline src/integrationTest/java/org/maproulette/client/IntegrationBase.java:105]: `this.getProjectAPI()`

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
  */
 public class IntegrationBase
 {
-    public static final String DEFAULT_PROJECT_NAME = "IntegrationTestProject";
+    public static final String DEFAULT_PROJECT_NAME = "IntegrationTestProject4";
```
Comment (mgcuthbert) [inline src/integrationTest/java/org/maproulette/client/IntegrationBase.java:16]: Why was the default project name changed?

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java [hunk 5]
```diff
@@ -40,7 +40,7 @@ public TaskAPI getTaskAPI()
     {
         if (this.taskAPI == null)
         {
-            this.taskAPI = new TaskAPI(this.getConfiguration());
+            this.taskAPI = new TaskAPI(this.getConfigurationExcludingProject());
```
Comment (mgcuthbert) [inline src/integrationTest/java/org/maproulette/client/IntegrationBase.java:43]: By not supplying the project name you are inadvertently changing the all the tests that use this. So unless there is a specific reason as to why you would retrieve the API without a project name this should be reverted.

File: src/main/java/org/maproulette/client/connection/MapRouletteConfiguration.java (hunk 1 above)
Comment (smaheshwaram) [reply src/main/java/org/maproulette/client/connection/MapRouletteConfiguration.java:36]: Sounds good. I can change that.

File: src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java (hunk 2 above)
Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java:67]: When I was using unit tests to compare object fields on both projects, I came across this test which is failing. It is failing because the geometry was not being updated. It needs to be addressed.

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java (hunk 3 above)
Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:105]: Need to push new changes. Modified it already.

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java (hunk 4 above)
Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:16]: When I was testing the same test cases couple of times, it was complaining that the project name exists already. Have to revert it and test it again to see what's going on. PR is WIP.

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java (hunk 5 above)
Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:43]: The existing tests will work with default constructor and with existing configuration. However, I am using these tests to verify with newly added construction which is been called from production tools.

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java (hunk 4 above)
Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:16]: forceDelete() in teardown() may not working as expected. Need to verify and fix it.

Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java]: When I am running existing tests from TaskAPIIntegrationTest, tests were failing saying that the default project name exists so unable to create a project with 500 error. I believe that the project name which was used before did n't get deleted using teardown() method. Currently, with newly modified project name, even though how many times I ran it, there are no issues because each time teardown() is force deleting the project which got created during the run.

File: src/integrationTest/java/org/maproulette/client/IntegrationBase.java (hunk 5 above)
Comment (danielduhh) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:43]: Instead of changing the existing tests, you can add news ones to cover the added constructor 
@smaheshwaram

Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/IntegrationBase.java:43]: I agree. I intended to do that.

File: src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java (hunk 2 above)
Comment (mgcuthbert) [reply src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java:83]: Why was it failing and why does resetGeometry resolve it? This test would have been previously succeeding.

Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java:83]: When I first ran all existing tests, every test except for this test was successful. I believe the issue was when adding GeoJson while updating existing test.
I have seen this below error when I ran it locally:
org.opentest4j.AssertionFailedError:
Expected :{“features”:[{“type”:“Feature”,“geometry”:{“type”:“Point”,“coordinates”:[3.1,4.2]},“properties”:{“name”:“UpdateGeometry”}}]}
Actual   :{“features”:[{“type”:“Feature”,“geometry”:{“type”:“Point”,“coordinates”:[1.1,2.2]},“properties”:{“name”:“TestGeometry”}}]}
For the first task creation, when we add geojson using this addGeoJson() method, it gets added in geometries of a task. During the update, when you try to add new geo json for new task it is not updating the geometries. It is still showing old geometries. That is why I tried to reset the geometries using method called resetGeometry() and tried to add new geoJson. Please let me know if it is clear to you or if you have alternate approaches to solve this.

Comment (mgcuthbert) [reply src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java:83]: So this test is only partly incorrect. Based on the code that is being executed (minus the resetGeometry that you added) the expected geometry should be 
```
{
“features”:[
{“type”:“Feature”,“geometry”:{“type”:“Point”,“coordinates”:[1.1,2.2]},“properties”:{“name”:“TestGeometry”}},
{“type”:“Feature”,“geometry”:{“type”:“Point”,“coordinates”:[3.1,4.2]},“properties”:{“name”:“UpdateGeometry”}}
]}
```
As on the updated task we are just called `addGeoJSON` which would just add the geojson to the current task not replace it. The reset function was built specifically for the situation where you wanted to replace the geometry, all it would also be nice to have a resetGeometryWith type of function.

Anyway all that to say, for this test we should probably test both scenarios, one where you are adding or appending geometry to the current geometry and make sure that it updates correctly and the other where we use the resetGeometry function and expect brand new geometry for the task.

A little outside the scope of this particular PR, but would be helpful to make sure that this works as expected in this scenario, and tested nicely.

Comment (smaheshwaram) [reply src/integrationTest/java/org/maproulette/client/api/TaskAPIIntegrationTest.java]: Done.

File: src/main/java/org/maproulette/client/model/Task.java [hunk 6]
```diff
@@ -187,6 +206,11 @@ private JsonNode buildGeometries() throws MapRouletteException
             result.set(TASK_FEATURES, generateTaskFeatures(this.points, this.geoJson));
             return result;
         }
+
+        public JsonNode getGeometries()
```
Comment (mgcuthbert) [inline src/main/java/org/maproulette/client/model/Task.java:210]: You shouldn't put a getter function in the builder. Why did you need this?

File: src/main/java/org/maproulette/client/model/Task.java [hunk 7]
```diff
@@ -100,24 +103,40 @@ public TaskBuilder resetGeometry()
...
                 {
-                    this.geometries(this.buildGeometries());
-                }
-            }
-            catch (final MapRouletteException e)
-            {
-                throw new MapRouletteRuntimeException(e);
+                    try
+                    {
+                        this.geometries(this.buildGeometries());
```
Comment (mgcuthbert) [inline src/main/java/org/maproulette/client/model/Task.java:132]: This shouldn't be changed. Plus all you are doing here is setting the geometries over and over again until you just set it to the last geoJson supplied. This function should remain as it was.

File: src/main/java/org/maproulette/client/model/Task.java (hunk 6 above)
Comment (smaheshwaram) [reply src/main/java/org/maproulette/client/model/Task.java:210]: I was trying to get geometries instance from updateTest(). I was using below line to get geometries of the previously created task

Comment (mgcuthbert) [reply src/main/java/org/maproulette/client/model/Task.java:210]: So once a task is built from the task builder, you can then retrieve the geometries and other properties of the task. But you shouldn't use the builder as a bean type of object. 

I think I am still confused one you are using the "taskBuilder" object and not the task itself to do what you are trying to do.

Comment (smaheshwaram) [reply src/main/java/org/maproulette/client/model/Task.java:210]: Yes, we can get geometries using Taskbuilder and well as Task class. But, the reason for my approach is all we are doing is trying to setGeometries for a task. So, since we are creating builder instance first, I thought of handling the geoJsons in builder itself. However, if you prefer to do it in Task then I can change code accordingly.

Comment (mgcuthbert) [reply src/main/java/org/maproulette/client/model/Task.java]: I am not sure I understand your reason, however the builder class is built for a specific purpose, and that is to build an object, in this case a Task. Once the Task is built then it can be used to retrieve information about it.

Comment (smaheshwaram) [reply src/main/java/org/maproulette/client/model/Task.java]: I understand. Reverted code in TaskBuilder and fixed the test. Please let me know if any questions. Thanks in advance.

**Comment by mgcuthbert** [comment]
please test this
